### PR TITLE
Relative Include path support in ZScript

### PR DESCRIPTION
**Pull request by RicardoLuis0**
Additional code in zcc_parser.cpp::DoParse for resolving relative include paths.
I've tested it some, seems to work so far.
If there is a better way to get the current file path or any other suggestions, I'll surely improve it.

['Code Submissions' Post](https://forum.zdoom.org/viewtopic.php?f=59&t=65777)

### CODE REVIEW

**Comment by RicardoLuis0** [comment]
Changing the code to require a './' in front of relative names, by suggestion from ZDoom discord users 3saster and phantombeta

**Comment by RicardoLuis0** [comment]
Changing from c strings to FString

**Comment by RicardoLuis0** [comment]
it helps, instead of having a wall of #includes you can spread it out, for example, have a "zscript/monsters/include.zs" file that includes all your monsters, and include them from there, but you still need to write the full path of, .ex "zscript/monsters/zombie/zombieVariant1.zs", with this you could simply write "./zombie/zombieVariant1.zs" for an #include in the "zscript/monsters" folder, and it doesn't seem to have any drawbacks

File: src/scripting/zscript/zcc_parser.cpp [hunk 1]
```diff
@@ -415,6 +415,24 @@ static void DoParse(int lumpnum)
 	for (unsigned i = 0; i < Includes.Size(); i++)
 	{
 		lumpnum = Wads.CheckNumForFullName(Includes[i], true);
+		if (lumpnum == -1 && Includes[i].IndexOf("./") == 0)//relative path resolving
+		{
+			FString fullPath = IncludeLocs[i].FileName.GetChars();//get full path, format 'wad:filepath/filename'
+			
+			size_t start = fullPath.IndexOf(":");//find first ':'
```
Comment (alexey-lysiuk) [inline src/scripting/zscript/zcc_parser.cpp]: The result of `FString::IndexOf()` is `long` (signed), not `size_t` (unsigned).

File: src/scripting/zscript/zcc_parser.cpp [hunk 2]
```diff
@@ -415,6 +415,24 @@ static void DoParse(int lumpnum)
...
+
+			size_t end = fullPath.LastIndexOf("/");//find last '/'
```
Comment (alexey-lysiuk) [inline src/scripting/zscript/zcc_parser.cpp]: Same for `FString::LastIndexOf()`.

File: src/scripting/zscript/zcc_parser.cpp [hunk 3]
```diff
@@ -415,6 +415,24 @@ static void DoParse(int lumpnum)
...
+
+			if(start!=-1&&end!=-1)
+			{
+				FString path = fullPath.Mid(start + 1, end - start);//extract filepath from string
+				path += Includes[i].Mid(2);//add relative path
+
+				lumpnum = Wads.CheckNumForFullName(path, true);//check for relative include
+				if (lumpnum == -1) IncludeLocs[i].Message(MSG_ERROR, "Include script lump %s not found", path.GetChars());
```
Comment (alexey-lysiuk) [inline src/scripting/zscript/zcc_parser.cpp]: This line is almost useless, and the same error will be printed twice with relative and full paths.

File: src/scripting/zscript/zcc_parser.cpp [hunk 4]
```diff
@@ -415,6 +415,24 @@ static void DoParse(int lumpnum)
...
 		lumpnum = Wads.CheckNumForFullName(Includes[i], true);
+		if (lumpnum == -1 && Includes[i].IndexOf("./") == 0)//relative path resolving
+		{
+			FString fullPath = IncludeLocs[i].FileName.GetChars();//get full path, format 'wad:filepath/filename'
+			
+			size_t start = fullPath.IndexOf(":");//find first ':'
+
+			size_t end = fullPath.LastIndexOf("/");//find last '/'
+
+			if(start!=-1&&end!=-1)
```
Comment (alexey-lysiuk) [inline src/scripting/zscript/zcc_parser.cpp]: Could you please use spaces? `if (start != -1 && end != -1)` is much easier to read.

**Comment by RicardoLuis0** [comment]
i'm on it, the changes will be comitted in a minute

**Comment by Iniquitatis** [comment]
Sure, this feature has its uses, but I still see it as a half-measure. In my opinion it could be much more convenient if ZScript had something like "#pragma once", or even a module system.

**Comment by alexey-lysiuk** [comment]
> Sure, this feature has its uses, but I still see it as a half-measure. In my opinion it could be much more convenient if ZScript had something like "#pragma once", or even a module system.

Will you code it?

**Comment by Iniquitatis** [comment]
Nope, sorry, I'm just an observer here.

**Comment by RicardoLuis0** [comment]
sorry, can you ask someone else to give their opinion?

**Comment by madame-rachelle** [comment]
I'm not too find of it requiring "./" to indicate a relative include path. In my opinion it should just search for the file relatively if it can't find it absolutely.

Obviously for compatibility purposes it should search the absolute paths first.

That's just my opinion, though... I don't think I should be a designated authority on this.

**Comment by RicardoLuis0** [comment]
> I'm not too find of it requiring "./" to indicate a relative include path. In my opinion it should just search for the file relatively if it can't find it absolutely.

It originally worked like that (relative if it doesn't find a file in absolute mode), but a bunch of people on discord said that i had to change it

**Comment by coelckers** [comment]
My main problem with changing it now is that directories are not isolated entities in a single mod. For the resource management /script/funstuff from a.zip and /script/funstuff are the same thing.
So a change to established behavior could easily pick up wrong files now.

That means, relative paths need a special designator so that the implementation won't create problems for existing mods.

**Comment by caligari87** [comment]
For what it's worth, I personally would vote for relative pathing with the standard designator (`./` `../` etc) to preserve existing behavior (paths without this would only ever be absolute), and to be explicit about what the author wants to do.

**Comment by RicardoLuis0** [comment]
support for '../' in relative includes added.

**Comment by coltongit** [comment]
@MajorCooke Would this be in good enough a state for QZDoom to pick up?

**Comment by MajorCooke** [comment]
No. It has a change request pending.

**Comment by alexey-lysiuk** [comment]
You can ignore my change request. While I’m still not sure that this feature is needed, you can merge to QZDoom if others think it’s useful.

**Comment by MajorCooke** [comment]
Can I get a small piece of example code so I know what to test?

**Comment by RicardoLuis0** [comment]
> Can I get a small piece of example code so I know what to test?

[rel_inc_test.zip](https://github.com/coelckers/gzdoom/files/3747265/rel_inc_test.zip)
simple test for the includes ( absolute, relative forwards, relative backwards )

**Comment by MajorCooke** [comment]
I've decided I'm going to hold off on this for now. Testing so many changes at once will mean that inevitably, catching up to GZDoom will fail on occasion from mixing too many. 

QZDoom is going to have to take these things slow and steady. We have enough on our plate at the moment.

**Comment by RicardoLuis0** [comment]
just updated to the last commit on master and fixed a minor conflict due to the file being moved, hope this can still move forward

**Comment by RicardoLuis0** [comment]
seems like it wasn't just a file rename but some deeper changes are needed too, so i'll be fixing it shortly

**Comment by RicardoLuis0** [comment]
Still no chance of this being added?

**Comment by alexey-lysiuk** [comment]
You’d better ask Graf about it. Usefulness of this feature is quite uncertain to me.

**Comment by RicardoLuis0** [comment]
@coelckers What's your view on adding this feature? In my opinion it can greatly improve the organization of projects, currently it adds relative include capability, for paths starting with './' or '../'

**Comment by madame-rachelle** [comment]
I do think this feature would be useful for transferring batches of ZScript files from one project to another. Also, most parsed languages worth their weight have such a feature.

**Comment by caligari87** [comment]
Just commenting to note that I've found a use where this would be helpful to have: I've received several requests to split my mod [Ugly as Sin"](https://github.com/caligari87/Ugly-as-Sin) to several smaller mods, so various features can be loaded separately. I've also received an equal number of pleas to leave it whole. I've settled on "both", allowing the mod to be loaded as a whole unit, or sub-folders to be split out and loaded as their own standalone mods. 

image

Relative pathing would make this trivial as the top-level combined mod could simply include the zscript file in the submodule...
```
version "4.5"
#include "ammopouch/zscript.zsc"
```

...which in turn would reference the submodule zscript as relative paths, so it could be loaded standalone. 
```
version "4.5"
#include "./module/ammopouch.zsc"
#include "./module/ammopouch_bulk.zsc"
#include "./module/ammopouch_hud.zsc"
```

As is, the top-level zscript file must include full paths to every submodule file.

This obviously isn't an insurmountable obstacle, but it is slightly annoying and relative paths would make it trivial to set up.

**Comment by RicardoLuis0** [comment]
just a squash, no code changes, still hoping for an approval for this PR

**Comment by coelckers** [comment]
Seems to work as intended.

**Pull request merged**